### PR TITLE
Support N-D tensors in Bilinear

### LINKED ISSUE #5601: I think Bilinear() should support 3D tensor like Linear() does
When I execute following code, it is gonna get 

> RuntimeError: matrices expected, got 3D, 2D tensors

`import torch.nn as nn`
`m = nn.Bilinear(2, 3, 4)`
`input1 = Variable(torch.randn(5, 3, 2))`
`input2 = Variable(torch.randn(5, 3, 3))`
`output = m(input1, input2)`
`print(output.size())`
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-23-b247cd5ad730> in <module>()
----> 1 output = m(input1, input2)
      2 print(output.size())

~/anaconda3/envs/deepchem/lib/python3.5/site-packages/torch/nn/modules/module.py in __call__(self, *input, **kwargs)
    323         for hook in self._forward_pre_hooks.values():
    324             hook(self, input)
--> 325         result = self.forward(*input, **kwargs)
    326         for hook in self._forward_hooks.values():
    327             hook_result = hook(self, input, result)

~/anaconda3/envs/deepchem/lib/python3.5/site-packages/torch/nn/modules/linear.py in forward(self, input1, input2)
    110 
    111     def forward(self, input1, input2):
--> 112         return F.bilinear(input1, input2, self.weight, self.bias)
    113 
    114     def __repr__(self):

~/anaconda3/envs/deepchem/lib/python3.5/site-packages/torch/nn/functional.py in bilinear(input1, input2, weight, bias)
    845         return Bilinear.apply(input1, input2, weight)
    846     else:
--> 847         return Bilinear.apply(input1, input2, weight, bias)
    848 
    849 

~/anaconda3/envs/deepchem/lib/python3.5/site-packages/torch/nn/_functions/linear.py in forward(ctx, input1, input2, weight, bias)
     15         # compute output scores:
     16         for k, w in enumerate(weight):
---> 17             torch.mm(input1, w, out=buff)
     18             buff.mul_(input2)
     19             torch.sum(buff, 1, keepdim=True, out=output.narrow(1, k, 1))

RuntimeError: matrices expected, got 3D, 2D tensors at /opt/conda/conda-bld/pytorch_1512383260527/work/torch/lib/TH/generic/THTensorMath.c:1411

### PR DESCRIPTION
**Pull request by li-roy**
Closes #5601.

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+
+namespace at { namespace native {
+
+Tensor bilinear(const Tensor& input1, const Tensor& input2, const Tensor& weight, const Tensor& bias) {
+  auto b_input1 = input1.unsqueeze(-2).unsqueeze(-2);
```
Comment (zou3519) [inline aten/src/ATen/native/Linear.cpp]: You should probably add some error checking for the input dimensions

File: torch/nn/modules/linear.py [hunk 2]
```diff
@@ -74,7 +74,10 @@ class Bilinear(Module):
 
     Shape:
         - Input: :math:`(N, \text{in1_features})`, :math:`(N, \text{in2_features})`
-        - Output: :math:`(N, \text{out_features})`
+          where :math:`*` means any number of additional dimensions. All but the last
```
Comment (zou3519) [inline torch/nn/modules/linear.py]: Did you mean `(N, *, in1_features)`? Ditto for in2 and output

File: aten/src/ATen/native/Linear.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  if (bias.defined() && bias.size(0) != weight.size(0)) {
+    throw std::runtime_error("Bias sizes does not match weight size");
+  }
+
+  auto b_input1 = input1.unsqueeze(-2).unsqueeze(-2);
+  auto b_input2 = input2.unsqueeze(-2).unsqueeze(-1);
+
+  auto output = at::matmul(at::matmul(b_input1, weight), b_input2);
+  output = output.squeeze(-1).squeeze(-2).sum(-1);
```
Comment (zou3519) [inline aten/src/ATen/native/Linear.cpp]: After the matmul output should be of size `(N, *, out, 1, 1)`. Why are you squeezing -1 then squeezing -2? shouldn't it be squeeze(-1).squeeze(-1)?

File: aten/src/ATen/native/Linear.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#include "ATen/ATen.h"
+#include "ATen/NativeFunctions.h"
+
+
+namespace at { namespace native {
+
+Tensor bilinear(const Tensor& input1, const Tensor& input2, const Tensor& weight, const Tensor& bias) {
+  if (input1.dim() != input2.dim()) {
+    throw std::runtime_error("Inputs should have the same number of dimensions");
```
Comment (zou3519) [inline aten/src/ATen/native/Linear.cpp]: Nit: these should be more descriptive in general. "Inputs should have same number of dimensions: got %d and %d".

File: aten/src/ATen/native/Linear.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  for (int64_t i = 0; i < input1.dim() - 1; i++) {
+    if (input1.size(i) != input2.size(i)) {
+      throw std::runtime_error("Batch dimensions of inputs do not match");
```
Comment (zou3519) [inline aten/src/ATen/native/Linear.cpp]: "Batch dimensions of inputs do not match in dimension %d: got %d and %d"

**Review by zou3519 — APPROVED** [review]
```
🥇          🥇 🥇 🥇 🥇    🥇 🥇 🥇       🥇 🥇🥇 🥇 🥇 
🥇          🥇                 🥇          🥇    🥇   🥇 
🥇          🥇     🥇 🥇      🥇           🥇    🥇   🥇 
🥇          🥇        🥇       🥇          🥇    🥇   🥇 
🥇 🥇 🥇    🥇 🥇 🥇 🥇      🥇          🥇    🥇   🥇 
```

**Comment by soumith** [comment]
I like richard's enthusiasm :D

**Pull request merged**